src/VisualQueryBuilder/QueryModellerBase.ts

```diff
@@ -0,0 +1,49 @@
+import { Registry } from '@grafana/data';
+
+import { BINARY_OPERATIONS_KEY, VisualQuery, VisualQueryBinary, QueryBuilderLabelFilter, QueryBuilderOperation, QueryBuilderOperationDef, VisualQueryModeller } from './types';
+
+export abstract class QueryModellerBase implements VisualQueryModeller {
+  protected operationsRegistry: Registry<QueryBuilderOperationDef>;
+  private categories: string[] = [];
+
+  constructor(getOperations: () => QueryBuilderOperationDef[]) {
+    this.operationsRegistry = new Registry<QueryBuilderOperationDef>(getOperations);
+  }
+
+  protected setOperationCategories(categories: string[]) {
+    this.categories = categories;
+  }
+
+  abstract renderOperations(queryString: string, operations: QueryBuilderOperation[]): string;
+
+  abstract renderBinaryQueries(queryString: string, binaryQueries?: Array<VisualQueryBinary<VisualQuery>>): string
+
+  abstract renderLabels(labels: QueryBuilderLabelFilter[]): string;
+
+  abstract renderQuery(query: VisualQuery, nested?: boolean): string
+
+  getOperationsForCategory(category: string) {
+    return this.operationsRegistry.list().filter((op) => op.category === category && !op.hideFromList);
+  }
+
+  getAlternativeOperations(key: string) {
+    return this.operationsRegistry.list().filter((op) => op.alternativesKey && op.alternativesKey === key);
+  }
+
+  getCategories() {
+    return this.categories;
+  }
+
+  getOperationDef(id: string): QueryBuilderOperationDef | undefined {
```

Comment:
Do you think we could use this opportunity to rename this to `getOperationDefinition`? I don't see a need to keep method names below a certain length and aids readability.

Reply:
Good point done in https://github.com/grafana/grafana-experimental/pull/98/commits/9af2866211c3a209af9a33de05f318521704a57b